subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ConfigurationResolver.java

```diff
@@ -25,18 +25,18 @@ public interface ConfigurationResolver {
     /**
      * Traverses enough of the graph to calculate the build dependencies of the given resolve context. All failures are packaged in the result.
      */
-    void resolveBuildDependencies(ResolveContext configuration, ResolverResults result);
+    ResolverResults resolveBuildDependencies(ResolveContext configuration);
 
     /**
      * Traverses the full dependency graph of the given resolve context. All failures are packaged in the result.
      */
-    void resolveGraph(ResolveContext resolveContext, ResolverResults results) throws ResolveException;
+    ResolverResults resolveGraph(ResolveContext resolveContext) throws ResolveException;
 
     /**
      * Calculates the artifacts to include in the result for the given resolve context. All failures are packaged in the result.
-     * Must be called using the same result instance as was passed to {@link #resolveGraph(ResolveContext, ResolverResults)}.
+     * Must be called using the same result instance that was returned by {@link #resolveGraph(ResolveContext)}.
```

Comment:
Yeah I agree that is the way to go here. In reality `ResolverResults` should be split up into multiple separate types, but I wanted to avoid toooo many changes in this PR given its size already. A follow-up to this would be to merge the concept of `ResolverResults` and the DefaultConfiguration's `ResolveState` into a single concept, while at the same type making sure each distinct state that a configuration's resolution could be in is represented as a separate type.